In [32]:
!pip install -q pyomo

In [33]:
from pyomo.environ import *

In [34]:
model= ConcreteModel()

In [35]:
#create variable
model.x1 =Var(domain=Reals)
model.x2 =Var(domain = Reals)
model.x3 =Var(domain = Reals)
model.x4 =Var(domain = Reals)
model.x5 =Var(domain = Reals)
model.x6 =Var(domain = Reals)
model.x7 =Var(domain = Reals)

In [36]:
#declare objective
model.cost = Objective(expr=  -3*model.x1 - 2*model.x2 + 8*model.x3 - 7*model.x4 + 0.19*model.x5 + 11*model.x6 + 20.2*model.x7, sense= minimize)

In [42]:
#declare constraints
model.con1 = Constraint(expr = 1*model.x1 + 1*model.x2 + 1*model.x3 + 1*model.x4 + 1*model.x5 + 1*model.x6 + 1*model.x7 == 256)
model.con2 = Constraint(expr = -1*model.x2 + model.x3 >= 10)
model.con3 = Constraint(expr = -3*model.x1 + 1*model.x5 -12*model.x6 <=105)
model.con4 = Constraint(expr = 18.5*model.x2 +30.5*model.x4 - 1*model.x7 >=28)
model.con5 = Constraint(expr = -1*model.x3 -4*model.x5 + 1*model.x7 >=31)
model.con6 = Constraint(expr = model.x1 >=-5)
model.con7 = Constraint(expr = model.x2 >=0)
model.con8 = Constraint(expr = model.x3 >=5)
model.con9 = Constraint(expr = model.x4 >=-3)
model.con10 = Constraint(expr = model.x5 >=-4)
model.con11 = Constraint(expr = model.x6 >=0)
model.con12 = Constraint(expr = model.x1 >=10)




This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [43]:
!apt-get install -y -qq glpk.utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155673 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [44]:
model.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x7 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fi

In [45]:
SolverFactory('glpk',excutable = '/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -950.760000000001
  Upper bound: -950.760000000001
  Number of objectives: 1
  Number of constraints: 13
  Number of variables: 8
  Number of nonzeros: 26
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010920524597167969
# ----------------------------------------------------------
#   Solution Information
# ----------

In [46]:
# display solution
print('\nCost = ', model.cost())

print('\nDecision Variables')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())
print('x6 = ', model.x6())
print('x7 = ', model.x7())




print('\nConstraints')
print('Constraint 1  = ', model.con1())
print('Constraint 2  = ', model.con2())
print('Constraint 3  = ', model.con3())
print('Constraint 4  = ', model.con4())
print('Constraint 5  = ', model.con5())
print('Constraint 6  = ', model.con6())
print('Constraint 7  = ', model.con7())
print('Constraint 8  = ', model.con8())
print('Constraint 9  = ', model.con9())
print('Constraint 10  = ', model.con10())
print('Constraint 11  = ', model.con11())
print('Constraint 12  = ', model.con12())



Cost =  -950.76

Decision Variables
x1 =  10.0
x2 =  0.0
x3 =  10.0
x4 =  215.0
x5 =  -4.0
x6 =  0.0
x7 =  25.0

Constraints
Constraint 1  =  256.0
Constraint 2  =  10.0
Constraint 3  =  -34.0
Constraint 4  =  6532.5
Constraint 5  =  31.0
Constraint 6  =  10.0
Constraint 7  =  0.0
Constraint 8  =  10.0
Constraint 9  =  215.0
Constraint 10  =  -4.0
Constraint 11  =  0.0
Constraint 12  =  10.0


**solution of 1.3**



**objective value = -950.76**

**value of decision var**

x1 = 10.0

x2 = 0

x3 = 10

x4 = 215

x5 = -4

x6 = 0

x7 = 25.0



**solution of 1.4**

x1+x2+x3+x4+x5+x6+x7=256 is active since x1+x2+x3+x4+x5+x6+x7 =256

-x2+x3 >=10 is active since -x2+x3=10

-3x1+x5-12x6<=105 is inactive since -3x1+x5-12x6= -34

18.5x2+30.5x4-x7>=28 is inactive since 18.5x2+30.5x4-x7 = 6532.5

-x3-4x5+x7 >= 31 is active since -x3-4x5+x7 = 31
